tests/Config/Plugin/Hook/ArrayMapStorageProvider.php

```diff
@@ -0,0 +1,190 @@
+{
+    public static function getFunctionIds(): array
+    {
+        return ['custom_array_map'];
+    }
+
+    public static function getFunctionStorage(FunctionDynamicStorageProviderEvent $event): ?FunctionStorage
+    {
+        $context = $event->getContext();
+        $statements_analyzer = $event->getStatementsAnalyzer();
+        $call_args = $event->getCallArgs();
+        $args_count = count($call_args);
+        $expected_callable_args_count = $args_count - 1;
+
+        if ($expected_callable_args_count < 1) {
+            return null;
+        }
+
+        $last_array_arg = $call_args[$args_count - 1];
+
+        if (ExpressionAnalyzer::analyze($statements_analyzer, $last_array_arg->value, $context) === false) {
+            return null;
+        }
+
+        $input_array_type = $statements_analyzer->node_data->getType($last_array_arg->value);
+
+        if (!$input_array_type) {
+            return null;
+        }
+
+        $input_value_type = self::getInputValueType($statements_analyzer->getCodebase(), $input_array_type);
+
+        $all_expected_callables = [
+            self::createExpectedCallable($input_value_type),
+            ...self::createRestCallables($expected_callable_args_count),
+        ];
+
+        $custom_array_map_storage = new FunctionStorage();
+        $custom_array_map_storage->cased_name = 'custom_array_map';
+        $custom_array_map_storage->template_types = self::createTemplates($expected_callable_args_count);
+        $custom_array_map_storage->return_type = self::getReturnType($all_expected_callables);
+
+        $input_array_param = new FunctionLikeParameter('input', false, $input_array_type);
+        $input_array_param->is_optional = false;
+
+        $custom_array_map_storage->setParams(
+            [
+                ...array_map(
+                    function (TCallable $expected, int $offset) {
+                        $param = new FunctionLikeParameter('fn' . $offset, false, new Union([$expected]));
+                        $param->is_optional = false;
+
+                        return $param;
+                    },
+                    $all_expected_callables,
+                    array_keys($all_expected_callables)
+                ),
+                $input_array_param
+            ]
+        );
+
+        return $custom_array_map_storage;
+    }
+
+    /**
+     * Resolve value type from array-like type:
+     *     list<int> -> int
+     *     list<int|string> -> int|string
+     */
+    private static function getInputValueType(Codebase $codebase, Union $array_like_type): Union
+    {
+        $input_template = self::createTemplate('TIn');
+
+        // Template type that will be inferred via TemplateInferredTypeReplacer
+        $value_type = new Union([$input_template]);
+
+        $templated_array = new Union([
+            new Type\Atomic\TArray([Type::getArrayKey(), $value_type])
+        ]);
+
+        $template_result = new TemplateResult(
```

Comment:
Same question here: most template methods being used here are in the internal section.
Maybe a good idea to create a wrapper class / function in the plugin or other namespace?

Reply:
The same. I don't know how it would be better to expose the internal API. Maybe @orklah can assist here.

Reply:
Personally I find the internal API pretty messy but I'm not sure either how to expose it in a way that would make sense. I guess we would start by listing every relevant resource to expose... It's been a while since I tweaked in plugins!